### PR TITLE
debug log: add log to indicate which upstream connection is attached to downstream

### LINKED ISSUE #27576: Adding correlation between downstream and upstream connection
After some recent debugging I noticed that there's no trace log that would be able to correlate between downstream connection and upstream connection. For debugging purposes, I believe it would be useful if there was some log similar to ``[C20] Attached to downstream connection [C19]``. I find it sometimes hard to manually figure out which upstream connection is associated with which downstream connection.

### PR DESCRIPTION
**Pull request by ohadvano**
Commit Message: connection pool: Add debug log to indicate which downstream connection is attached
Additional Description: Resolves #27576, the idea is adding a debug log to help understand to which downstream connection an upstream connection is attached to. Opening to initial review if the approach of adding ``connection_id`` to the ``AttachContext`` makes sense.
Risk Level: Medium
Testing:
Docs Changes: None
Release Notes: None
Platform Specific Features: None

### CODE REVIEW

File: source/common/conn_pool/conn_pool_base.h
Comment (ohadvano) [inline source/common/conn_pool/conn_pool_base.h:26]: Adding connection id here means that it would need to propagate through ``HttpAttachContext`` or ``TcpAttachContext``, when calling ``newStream``, so a change to ``newStream`` signature is required with a new paramater:
https://github.com/envoyproxy/envoy/blob/a9571872dbf1c3c125ab0aaba086e5416e1048f9/envoy/http/conn_pool.h#L114

Alternatively, ConnPoolImplBase could be change so it also gets the connection id in the ctor:
https://github.com/envoyproxy/envoy/blob/a9571872dbf1c3c125ab0aaba086e5416e1048f9/source/common/conn_pool/conn_pool_base.h#L186

Will appreciate guidance if any of these approaches makes sense

**Review by wbpcode — COMMENTED** [review]
Thanks for your contribution. But I think not all upstream streams have related downstream connections?

I personally prefer implement this feature in the `router` filter, then we needn't update the code of connection pool.

**Comment by ohadvano** [comment]
Thanks @wbpcode, 

> Thanks for your contribution. But I think not all upstream streams have related downstream connections?

In which case for example there would be an upstream connection that's not related to a downstream connection?

> I personally prefer implement this feature in the router filter, then we needn't update the code of connection pool.

It seems to me like implementing this in the router would only take care of HTTP case but implementing this in the connection pool code would also handle raw TCP upstreams.

**Comment by wbpcode** [comment]
> In which case for example there would be an upstream connection that's not related to a downstream connection?

All request/streams that created by the envoy directly. For example, upstream connection for xds request.

**Comment by wbpcode** [comment]
> It seems to me like implementing this in the router would only take care of HTTP case but implementing this in the connection pool code would also handle raw TCP upstreams.

I think it's only one line log code in the router filter? It's ok for me to update both router filter and tcp proxy.

**Comment by ohadvano** [comment]
> All request/streams that created by the envoy directly. For example, upstream connection for xds request.

I see. I'm wondering what would be the ``AttachContext`` object in that case, because I could only find ``HttpAttachContext`` and ``TcpAttachContext``. What would be the attach context object for Envoy self managed upstream connections?

**Comment by ohadvano** [comment]
ping @wbpcode, @zuercher

File: source/common/tcp_proxy/upstream.cc
Comment (wbpcode) [inline source/common/tcp_proxy/upstream.cc:207]: why cannot do this in the `onGenericPoolReady` of `TcpProxy::Filter`?

Comment (ohadvano) [reply source/common/tcp_proxy/upstream.cc:207]: In ``TcpConnPool`` case it works, but for ``HttpConnPool`` the behavior is slightly different - first, ``onPoolReady`` is called to indicate network connection is ready, then it sends the ``CONNECT`` headers, then if there's a headers response ``onGenericPoolReady`` will be called. Since the attach of upstream connection to downstream connection does not depend on L7 data, I logged it at ``onPoolReady`` as it's an earlier stage.

**Review by wbpcode — APPROVED** [review]
LGTM. Thanks.

/assign-from @envoyproxy/senior-maintainers for second review or merge.

**Review by zuercher — APPROVED** [review]

**Pull request merged**
